[Step3] 날씨 정보(전) 타코캣(tacocat)

Ldoy opened the pull request:
# 날씨정보 Step3 PR

안녕하세요 찌루루~~~~~ @jae57 기존 일정보다 많이 늦어버렸네요..🥲 늦은만큼 기능을 최대한 구현해 보려 했으나 구현하지 못한 기능이 산더미 입니다😵 하지만 이번 프로젝트는 PR 일정을 맞추는 것 또한 프로젝트의 요구사항 중 하나라고 생각하여 PR드립니다😭

더불어 지난 번 PR에서 말씀 하셨던 부분 빠지지 않고 구현하려했는데 막상 결과물을 보니 많이 반영하지 못한 것 같습니다. 여러가지 고민해서 주셨던 피드백인데 혹시 제가 놓친 부분이 있다면 번거로우시더라도 한 번만 더 짚어주시면 감사드리겠습니다. 🙇‍♀️

# 앱 동작
<img src = "https://i.imgur.com/caggOd1.gif" width = 200, height = 400>


# 고민한 부분

#### 1. 뷰컨의 역할을 줄일 수 있는 방법
- 첫 번째 시도 : `MVVM` 패턴으로 변경
    - 결과 : PR일정을 맞추면서 진행하기엔 무리가 있닫고 판단하였습니다. 더불어 ViewModel을 구현하기 위해 많은 코드가 추가되더라구요. 그래서 이 패턴은 다음 기회에 한번 더 도전해 보려합니다. 
    
- 두 번째 시도 : 뷰컨의 역할을 최대한 줄이기
    - 방법
        - 보통 뷰컨이 역할을 담당하는TableViewDataSource, URLSessionDataDelegate 를 커스텀 객체로 따로 만들어 분리하였습니다.
        - CLLocationManagerDelegate의 역할은 따로 분리하지 않았습니다. 그 이유는 사용자가 위치정보를 거부하는 경우 알람을 띄울 수 있도록 하기 위해서 입니다. (더불어 뷰컨이 역할 하나정도 하는건 괜찮지 않을까🤔 하는 생각도 들었습니다)
        
        - 하지만 PR을 쓰면서 다시 생각해보니 해당 CLLocationManagerDelegate도 분리하고 Notification를 통해 알람을 띄울 수 있도록 해도 좋았을 것 같습니다. 

#### 2. 비동기적으로 얻게되는 위치정보, 날씨정보, 날씨아이콘이미지의 요청 흐름의 관하여 

- 전체적인 흐름
` 위치정보 요청 -> 위치정보 데이터 받아옴 -> 5일날씨예보정보를 요청하는 객체에 전달 -> 데이터디코딩 -> 날씨아이콘이미지 요청하는 객체에 전달`

- `위치정보`
    - 뷰컨의 viewDidLoad에서 locationService를 유저에게 요청한 후 유저가 accept하게 되면 위치정보를 받아올 수 있도록 하고 있습니다. 해당 위치 정보는 
`OpenWeatherSessionDelegate`로 전달하고 있습니다. 
- `5일날씨예보정보`
    - 위치정보 요청 후 완료된 데이터가 들어오는 곳에서 `WeatherDataHolder` 라는 singleton객체로 전달하였습니다. 
    - `WeatherDataHolder` 객체에선 디코딩한 `5일날씨예보정보`를 가지고 있으며 추후 `5일날씨예보정보`와 `현재날씨정보`의 구분은 OpenWatherAPI 객체의 urlPath정보를 통해 분기할 수 있도록 하였습니다.

    - `WeatherDataHolder` 를 singleton 패턴으로 구현한 이유
        - 현재 각각의 역할을 하는 객체를 따로 구분하였기 때문에 서버 데이터 요청이 확실히 완료된 시점(= 데이터가 있음을 확신할 수 있는 시점)에서 tableViewDatasource에 데이터를 전달할 수 있는 방법이라고 생각했습니다.
        - 또한 TableView의 HeaderView에 현재날씨에 대한 정보를 전달 해 주어야 하는데 이 시점이 `5일날씨예보` 정보가 들어오는 때와 다를 것 같다는 생각이 들었습니다. 

- `날씨아이콘이미지`
    - 현재 날씨 아이콘 이미지를 서버에 요청하기 위해선 `5일날씨예보정보`를 디코딩한 후 URL에 들어갈 path 정보를 가져와야합니다. 
    - `tableView(_:cellForRowAt:)`  메소드에서 image를 서버에 요청한 이유
        - 테이블뷰가 리로드 되는 시점은 `5일날씨예보정보` 가 있음을 확신할 수 있고(Notification을 보내도록 구현했기 때문) 따라서 이 메소드 내부에서 ImgaeLoader객체를 통해 이미지를 요청하였을 때 원하는 데이터를 얻을 수 있을 것이라고 생각했습니다. 
    - `ImamgeLocader`를 싱글턴으로 구현 한 이유
        - 현재 데이터 소스는 각 cell을 만들 때 마다 호출 될 것인데 그 때마다 인스턴스를 만드는 과정이 들어가는 것 보단 타입메서드를 통해 한 번 메모리에 올라간 후 계속 사용하는게 더 효율적일 것 같다는 생각이 들었습니다. 
        - 더불어 refresh contorl 을 통해 tableView를 다시 그릴 수 있는 기능이 있기 때문에 요청이 자주 이루어질 것 같다고 생각했습니다. 
        

#### 3. request만드는 메소드가 있으니 최대한 활용하기 
- 기존에 url로 dataTask를 만드는 것 대신 만들어 둔  request를 전달 할 수 있도록 구현하였습니다. 
    - 변경 전 
    ```swift
    func request(_ route: EndPointType, _ session: URLSession) {
        let request = self.buildRequest(from: route)
        
        guard let url = request.url else {
            return
        }
        
        task = session.dataTask(with: url)

        self.task?.resume()
    }
    
    private func buildRequest(from route: EndPointType) -> URLRequest {
        var request = URLRequest(url: route.baseUrl)
        
        switch route.httpTask {
        case .request(withUrlParameters: let urlParameter):
            self.configureRequestUrl(&request, urlParameter)
        }
        
        return request
    }
    
    private func configureRequestUrl(_ request: inout URLRequest, _ urlParameters: Parameters) {
        do {
            try URLManager.configure(urlRequest: &request, with: urlParameters)
        } catch {
            print(error)
        }
    }
    ```
    
    - 변경 후 코드 
    ```swift
    private func buildRequest(from route: EndPointType) -> URLRequest {
        var request = URLRequest(url: route.baseUrl)
        request.url?.appendPathComponent(route.path.rawValue, isDirectory: false)
        
        switch route.httpTask {
        case .request(withUrlParameters: let urlParameter):
            do {
                try URLManager.configure(urlRequest: &request, with: urlParameter)
            } catch {
                print(error)
            }
        }
        return request
    }
    ```

#### 4. 테이블뷰의 lazyLoading 문제 해결하기 
- 아이콘의 이미지를 요청하는 ~~`ImgaeLoader` 객체를 싱글턴으로 만들어서~~  `ImageCacher`객체를 통해 해결하였습니다. 

#### 5. 명확한 타입으로 날씨정보 데이터를 가지고 있도록 구현하기 
- `OpenWeatherSessionDelegate`객체에서 데이터를 받아오면 `WeatherDataHolder`객체에서 해당 데이터를 명확한 타입으로 가지고 있을 수 있도록 구현하였습니다. 

#### 6. URLSessionDataDelegate에 관하여
- 지난 번 PR에서 마지막으로 코멘트 주신 부분에 대해 고민 해 보았습니다. 

    > URLSession 의 delegate 는 weak 으로 선언되어있는지 한번 확인해보시고,현재처럼 ViewController 가 URLSessionDelegate 를 채택하게 되면 생길 수 있는 문제가 무엇이 있을지 생각해보시면 좋을 거 같아요!
    
    - 생길 수 있는 문제 : Viewcontroller의 역할이 너무 많아진다. 매번 session을 만들어 주어야 하기 때문에 URLSession.shared에서 제공하는 기능 외의 것을 사용하지 않는 경우 별다른 장점을 얻을 수 없다. 
    - URLSession의 delegate가 weak로 선언되어있는지의 여부 : strong하게 선언되어있음. 따라서 session을 다쓰고 나선 finishTasksAndInvalidate() 등의 메서드를 호출 해 주어야 함. 
    > delegate
A session delegate object that handles requests for authentication and other session-related events.
This delegate object is responsible for handling authentication challenges, for making caching decisions, and for handling other session-related events. If nil, the class should be used only with methods that take completion handlers.

    > **Important**
The session object keeps a strong reference to the delegate until your app exits or explicitly invalidates the session. If you do not invalidate the session by calling the invalidateAndCancel() or finishTasksAndInvalidate() method, your app leaks memory until it exits.

# 질문 드리고 싶은 부분
1. tableView lazyLoading문제를 어느정도 해결했다고 볼 수 있을지 궁금합니다. singleton 패턴을 이용해서 해결한 방향에 대해 어떻게 생각하시나요? 

3. `DispatchQueueGroup`, `DispatchQueueWorkItem` 을 이용해서 조금 더 네트워크 요청을 간단하게 할 수 있는 부분이 있을지 궁금합니다. 제가 `위의 두 객체의 사용이 익숙치 않아서 사용흐름이 보이지 않는건가..🥲` 하는 생각이 들어 질문 드립니다 .

5. 이후 스텝에서 Notificationcenter를 이용해서 정보나 알림을 전달하는 부분을 많이 사용하게 될 것 같은데요. 현업에선 이런 방법을 많이 사용하는지 궁금합니다. 
6. URLSession을 invalid하는 위치에 관하여 
    - session 을 invalid하게 만드는 시점은 언제가 좋을까요? 현재는 refresh control을 구현하지 않았기 때문에 데이터를 받아온 후 WeatherDataHolder를 만든느 부분에서 finishTasksAndInvalidate() 를 호출하였는데 아마 위치가 바뀌어야 하지 않을까 하는 생각이 들었습니다. 
        - 이유 : 한 번 session 이 invalid되면 다시 session을 만들어야 한다. 하지만 현재 sessionDelegate인스턴스가 만들어지는 부분은 OpenWeatherMainViewController이 초기화 되는 부분이므로 앱이 한 번 실행 된 이상 다시 sessionDelegate를 만들지 않기 때문.
 
4. Grouping에 관하여
    - 이전에는 타입들을 모두 개별적인 file에 두고 Grouping을 했었는데요 그렇게 하다보니 코드를 이해하는데 너무 어렵더라구요. 그래서 이번엔 관련된 타입들이나 extension의 경우 같은 file에서 구현하도록 하였습니다. 이렇게 하는게 찌루루가 느끼시기엔 코드의 가독성 부분에 도움이 되는지 궁금합니다. 


# 리팩토링 할 부분
1. 네트워크객체를 독립적으로 만들기 
   -  지난 번 PR때 피드백 주신 부분
    > 우선, 현재 NetworkManager 의 역할이 무엇이고, Router 의 역할이 무엇인지 확실하게 할 필요가 있을 거 같아요! 현재 task 를 구성하는 부분을 NetworkManager 로 옮겨보는 건 어떨까요? 
    - 처음 `Router`를 구현 했던 이유가 날씨정보요청 뿐 아니라 지금처럼 날씨아이콘 이미지 요청등의 네트워크 요청에서도 사용하기 위해서였는데 재사용을 전혀 못 하고 있습니다. 말씀하신 것 처럼 Router의 역할이 명확하지 않은 상태입니다.🥲 그래서 이미지 요청 하는 과정에서 Router를 사용할 수 있도록 시도해 본 후 NetworkManager객체로의 역할 이동을 시도하려고 합니다. 
    
3. ApiModeling 프로토콜에 관하여 
    - 해당 프로토콜은 URLSessionDataDelegate객체에서 API를 만들기 위해 급하게(?) 만든 타입입니다. 전체적인 네트워크 요청의 흐름에서 해당 프로토콜이 가지고 있는 메소드를 위치시킬 곳을 고민 해 보겠습니다. 
4. 효과적인 오류처리 
    - 현재 대부분의 error가 발생하는 곳에서 print를 하고 있습니다. Notification을 통해 추후 alert을 할 수 있도록 구현 해 보려 합니다. 
5. 전체적인 NetworkManger 내부의 메소드와 역할에 관하여 
    - EndPoint 타입을 잘 활용하여 URL을 만드는 역할과 흐름, api와 request를 만드는 역할과 흐름을 다시 정리해 보려합니다. (지금은 이곳 저곳 위치한 것 처럼 느껴지더라구요..)

# 구현하지 못 한 부분
1. 헤더뷰
2. cell 내부 요소들의 정확한 위치 
3. 날짜형식
    - 날짜 포맷을 프로젝트 기술서 있는대로 바꾸고 싶었는데 완벽하게 바꾸지 못했습니다. 
5. refresh control

Ldoy committed:
refactor: api키 난독화 할 필요없기 때문에 삭제 후 string 타입의 상수로 저장

Ldoy committed:
refactor: 필요없어진 APIError 열거형 삭제

Ldoy committed:
refactor: Router타입에서 사용하지 않는 data 속성 삭제

Ldoy committed:
refactor: 가독성 위해 뷰컨 내부 속성 위치 변경

Ldoy committed:
refactor: MVVM에 맞춰 그룹 위치 재선정

Ldoy committed:
feat: 위치정보를 표시한 테이블 뷰 추가

Ldoy committed:
feat: 테이블뷰 셀 ViewModel 구현

Ldoy committed:
feat: custom cell  구현

Ldoy committed:
feat:  MVVM모델로 바꾸는 중(ing)

- 뷰컨의 로직을 뷰모델로 옮기기

Ldoy committed:
feat: 테이블뷰 셀을 위한 viewModel 구현

- LocationManager 에 필요한 것만 import 진행

Ldoy committed:
feat: 날씨 아이콘 대체할 수 있는 placeHolder용 이미지 추가

- NetworkManager파일을 WeatherNetworkManager 이름으로 변경

Ldoy committed:
feat: EndPoint에 url경로 추가하기 위한 Path 타입 추가

Ldoy committed:
feat: custom cell 구현 및 cell에 데이터 전달하는 CellHolder구현

Ldoy committed:
refactor: NetworkManger타입 에서 디코딩 한 정보를 전달

- 디코딩한 데이터 전달하기 위한 Holder타입 구현
- 인스턴스 생성 없이 apiKey접근하기 위해 static으로 변경
- fetchData메소드에서 네트워크 요청 결과를 Holder에 담아 decoding된 상태로 전달

Ldoy committed:
feat: MVVM패턴 에서 Observer 역할을 하는 Box<T> 구현

Ldoy committed:
feat: HTTP응답코드 확인하는 메서드 추가

- route 매게변수의 path속성을 url에 추가하는 부분 구현

Ldoy committed:
feat: viewmodel이용해 cell 구성하도록 함

Ldoy committed:
refactor: MVC 구조에 사용하지 않는 타입삭제 및 이름 변경

Ldoy committed:
refactor: ViewController에서 OpenWeatherMainViewController 로 이름변경

- ApiModeling 주석 제거
-

Ldoy committed:
feat: 사용하지 않는 타입 및 메서드 삭제

- Holder타입 및 checkStatusCode메서드

Ldoy committed:
refactor: tableView가 뷰에 그려지도록 로직 수정

- cell의 stackView의 오토레이아웃 구현한 부분에서 translatesAutoresizingMaskIntoConstraints 속성 false로 지정

Ldoy committed:
feat: OpenWeatherSessionDelegate에 받아온 데이터 저장하는 속성 추가

Ldoy committed:
feat: sessionDelegate타입에서 데이터 전달하기 위한 Holder타입구현

Ldoy committed:
feat: UIView 에 오토레이아웃 구현하는 extension메소드 추가

- Holder 구조체 싱글턴으로 구현
- ApiModeling 내부 메소드 매게변수 이름 오타 수정

Ldoy committed:
feat: 날씨아이콘 이미지 로딩 및 캐싱하기 위한 타입 구현

Ldoy committed:
feat: 네트워크 요청을 통한 날씨정보를 받은 후 테이블뷰 reload하기 위한 notification 구현

Ldoy committed:
feat: tableViewDataSource에서 cell로 정보 전달하는 로직 구현

Ldoy committed:
refactor: 아이콘 이미지가 찌그려저서 들어가는 현상 해결

- tableView의 cell 높이 명시적으로 할당

Ldoy committed:
refactor: plist에 사용자위치정보 요구하는 이유 조금 더 상세하게 변경

Ldoy committed:
refactor: 날씨정보를 담고있는 holder에 맞는 이름으로 수정

Ldoy committed:
chore: 오타수정 및 사용하지 않는 변수 제거

Ldoy committed:
feat: session의 delegate가 strong하게 선언되어있기 때문에 finishTasksAndInvalidate 메서드 실행

Ldoy committed:
feat: 뷰컨에서 속성으로 가지지 않아도 되는 부분 CllocationManagerDelegate메서드 내부로 위치 변경

jae57 reviewed (commented):
@Ldoy
안녕하세요! 타코캣 🙋‍♂️

PR 이 확실히 지난번보다 자세해졌네요! 모두 꼼꼼히 읽어보았습니다! 
전반적으로 지난번 코드리뷰에 남긴 부분들 모두 신경써주셨네요!
SessionDelegate 에 관하여 추가적으로 남긴 내용도 잘 찾아주셨고, buildRequest 메서드 리팩토링도 깔끔하게 잘 정리해주셨네요!
(자세한 PR 덕에 구현하신 이유들이 명확해져서 저는 몹시 좋은데, PR 보내실 때 오래걸리실까 염려되네요..! 
부담은 갖지마시고, 편한 마음으로 작성해주세요! 😂)

이미지 지연 로딩 문제 없이 5일 간의 날씨 데이터 목록 잘 보이는 것 확인하였습니다. 
다만, 구현하지 못 한 부분에 써주신 것 처럼 이번 스텝에 핵심이 되는 부분들 구현이 아쉽네요 🥲
남긴 코멘트들은 참고만 하시고, **우선 구현에 집중해봅시다!**

**질문에 대한 답변**
1
테이블뷰의 이미지 지연로딩 문제가 해결된 것은 캐시를 구현했기 때문이지 싱글턴 패턴을 이용했기 때문은 아니라고 생각합니다.
( 고민한 부분 4번에 설명해주신 것은 ImageCacher 를 ImageLoader 라고 잘 못 쓴거겠죠? )

2
이제 써보시려는 건가요? 어떤 부분이 복잡하다고 느끼셨나요? 
(질문이 좀 더 구체화된다면 더 만족스러운 답을 받으실 수 있을 거 같아요!)

DispatchQueueGroup 과 DispatchQueueWorkItem 은 비동기 작업들이 많이 존재할 때, 그들간의 우선순위를 지정해주거나 세밀한 시점들을 노티받는 등 간단하게 사용할 수 있다기보다 
좀 더 세밀한 터치(?)를 해주고 싶을 때 사용한다고 생각합니다. 

3
개인적으로는 꼭 필요한 경우가 아니라면 ( 다수의 객체에 알려줘야 하는 경우 ) 주로 Delegate 를 이용하는 편입니다.
( 좀 더 명확하게 역할을 표현하고 지정해준다는 느낌이 들어서요.🧐 Notification Center 의 경우 코드들이 왠지 모르게 흩뿌려져 있는 느낌이라, 
다시말하면 노티를 보내는 쪽과 받는 쪽을 한번에 찾아내기 어렵기 때문에, 둘 중 한 쪽에 수정이 가해진다면 이로인한 사이드이펙트를 막기 위해 
노티를 쓰는 모든 코드들을 살펴봐야 한다는 것 때문에 피하는 편입니다 )

4
먼저 지난번 리뷰에 추가로 남긴 부분 관련 내용 잘 찾아주셨네요! 👏
조금 덧붙이자면, step2 때의 코드를 기준으로 보면, session 과 view controller 가 서로를 strong 하게 
잡고있기 때문에 순환 참조 문제가 발생합니다. 아마 해당 view controller 는 앱이 종료되기 전까지 메모리에서 해제되지 않을거에요.
만약 viewController 가 URLSessionDelegate 를 채택하는 상황을 그대로 가져간다고 한다면 
session 을 적절한 시기에 invalid 시켜주어야 겠지만
( 매 통신마다 session을 invalid 하고, 다시 만드는 것도 비용이 드니 좋은 방법은 아닐 것 같네요! ) 
요번 PR 에서는 OpenWeatherSessionDelegate 라는 객체가 session 을 소유하고, URLSessionDelegate 를 채택하게 되었으니 viewController 는 적절한 시기에 메모리에서 잘 해제될 겁니다. 

5
분량이 길지만 않다면 코드 이해 관점에서 바로 볼 수 있도록 한 파일에 묶는 것도 좋은 방법이라고 생각합니다.
개인적으로 Service 폴더가 조금 정리되어야 할 것 같다는 생각이 드네요! (이를테면, NetworkLayer 안에 Networking 이 또 있다든지, 이름에 Manager 가 붙는 것들은 Service 객체에 해당한다고 생각하는데 여기저기 흩어져 있다든지 )
도메인을 기준으로 Grouping 을 할 것인지, 역할을 기준으로 Grouping 을 할 것인지 이 두가지 기준이 함께 존재하는 거 같아요
둘 중 하나 기준점을 잡으시고, 그에 따라 정리해보시면 좋겠습니다!

Ldoy committed:
refactor: 테이블뷰 셀 이름 오타 수정

- FiveDaysForcecastCell 에서 FiveDaysForecastCell 으로 변경

Ldoy committed:
refactor : 동사로 된 파일명 명사로 수정

- WeatherInfoDecode 에서 WeatherInfo로 변경

Ldoy commented on WeatherForecast/WeatherForecast/Models/Decoding/WeatherInfoDecode.swift:
```diff
@@ -26,6 +26,33 @@ struct Parser {
     }
 }
 
```
이런.. 기본적인 부분을 놓쳤네요.. 넵~! 수정하였습니다!

Ldoy committed:
refactor : 타입이름 수정 및 직관적인 그룹핑

- OpenWeatherApi 에서 OpenWeatherAPI로 변경
- UIView 를 extension하는 것을 구현한 파일을 Extension그룹으로 위치변경

Ldoy commented:
찌루루 상세한 답변 감사합니다~~🤩🤩🤩 피드백 주신 부분 반영하였습니다🌟 더불어 날씨정보(전)Step3 나머지 기능과 더불어 날씨정보(후)Step1 내용을 같이 구현하여 PR 드려도 될지 여쭤보고 싶습니다~~

> 1. 테이블뷰의 이미지 지연로딩 문제가 해결된 것은 캐시를 구현했기 때문이지 싱글턴 패턴을 이용했기 때문은 아니라고 생각합니다.( 고민한 부분 4번에 설명해주신 것은 ImageCacher 를 ImageLoader 라고 잘 못 쓴거겠죠? )

맞습니다!!

> 2 이제 써보시려는 건가요? 어떤 부분이 복잡하다고 느끼셨나요? (질문이 좀 더 구체화된다면 더 만족스러운 답을 받으실 수 있을 거 같아요!)DispatchQueueGroup 과 DispatchQueueWorkItem 은 비동기 작업들이 많이 존재할 때, 그들간의 우선순위를 지정해주거나 세밀한 시점들을 노티받는 등 간단하게 사용할 수 있다기보다 좀 더 세밀한 터치(?)를 해주고 싶을 때 사용한다고 생각합니다.

아하!!! 사실 프로젝트 내 비동기적인 요청들의 결과를 받는 시점을 Notification을 통해서 진행하는 것 보단  DispatchQueueWorkItem과 group을 통해 진행하고싶었는데 잘 보이지 않더라구요. 뷰컨의 아래 메소드에서 해 보고 싶었는데 다시 생각해보니 적절하지 않을 것 같습니다. (group 은 보통 비동기적인 작업 내부에서 비동기적인 작업을 처리하고 그 완료시접을 알고 싶을 때 사용하는 것으로 이해했는데 현재의 흐름과는 맞지 않은 것 같다고 생각했습니다. )
```swift
   func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let latitude = locations.last?.coordinate.latitude,
              let longitude = locations.last?.coordinate.longitude else {
            return
        }
        
        let sessionDelegate = OpenWeatherSessionDelegate()
        let networkManager = WeatherNetworkManager()
        
        guard let api = networkManager.buildApi(weatherOrCurrent: .forecast, location: (latitude, longitude)) else {
            return
        }
        
        networkManager.fetchOpenWeatherData(requiredApi: api, sessionDelegate.session)
    }
```

더불어 이번 프로젝트 기간 내에 배운 활동학습 주제 중 하나가 DispatchQueueGroup과 DispatchQueueWorkItem 여서 `이 타입들을 이용해 조금 더 간단하게 정리할 수 있는 부분이 있는 건가? 내가 뭘 놓치고 있는건가` 하는 생각이 들어서 질문드렸습니다ㅎㅎㅎ 

> 4
먼저 지난번 리뷰에 추가로 남긴 부분 관련 내용 잘 찾아주셨네요! 👏
조금 덧붙이자면, step2 때의 코드를 기준으로 보면, session 과 view controller 가 서로를 strong 하게
잡고있기 때문에 순환 참조 문제가 발생합니다. 아마 해당 view controller 는 앱이 종료되기 전까지 메모리에서 해제되지 않을거에요.
만약 viewController 가 URLSessionDelegate 를 채택하는 상황을 그대로 가져간다고 한다면
session 을 적절한 시기에 invalid 시켜주어야 겠지만
( 매 통신마다 session을 invalid 하고, 다시 만드는 것도 비용이 드니 좋은 방법은 아닐 것 같네요! )
요번 PR 에서는 OpenWeatherSessionDelegate 라는 객체가 session 을 소유하고, URLSessionDelegate 를 채택하게 되었으니 viewController 는 적절한 시기에 메모리에서 잘 해제될 겁니다.

아하!! 그렇군요!!! 그럼 현 상황에선 invalid 진행하지 않아도 되겠네요~~ㅎㅎ 감사합니다~~~

> 5
분량이 길지만 않다면 코드 이해 관점에서 바로 볼 수 있도록 한 파일에 묶는 것도 좋은 방법이라고 생각합니다.
개인적으로 Service 폴더가 조금 정리되어야 할 것 같다는 생각이 드네요! (이를테면, NetworkLayer 안에 Networking 이 또 있다든지, 이름에 Manager 가 붙는 것들은 Service 객체에 해당한다고 생각하는데 여기저기 흩어져 있다든지 )
도메인을 기준으로 Grouping 을 할 것인지, 역할을 기준으로 Grouping 을 할 것인지 이 두가지 기준이 함께 존재하는 거 같아요
둘 중 하나 기준점을 잡으시고, 그에 따라 정리해보시면 좋겠습니다!

아하 도메인과 역할에 대한 기준점이 존재하는군요!! 감사합니다~~!!! `도메인`...이 무엇인지 모르는 상태라 우선  Service 폴더 정리와 더불어 역할을 기준으로 grouping 해 보겠습니다~~~!!ㅎㅎㅎ 기능구현 마치면 도메인에 대해 공부해 보겠습니다!!🥳

Ldoy committed:
refactor: 역할에 맞도록 그룹핑 진행

- ReqeustConfigurable 프로토콜 제거 및 URLConfigureManager를 Manager그룹에 추가
- Error 그룹 생성 및 NetworkError파일 추가

Ldoy committed:
refactor: HTTPTask타입 대신 Path, Query 요소 넣을 수 있도록 구현

Ldoy committed:
refactor: NetworkManager, Router의 역할 정리 및 캡슐화

Ldoy committed:
feat: 네트워크 객체 타입 구현 완료

Ldoy committed:
refactor: icon이미지 서버요청하는 객체는 따로 두는 방향으로구현

- 서버요청하는 객체에 추가했던 날씨 아이콘 관련 case 제거

Ldoy committed:
feat: 테이블뷰 헤더뷰의 기본적인 요소 구현

- 아직 헤더뷰 내부 오토레이아웃 및 네트워크 요청 흐름 구현하지 않음

Ldoy committed:
feat: 헤더뷰 내부 컨텐츠 구성하는 configure메소드 구현

- 테이블뷰 델리게이트 구현
- 테이블뷰와 헤더뷰 identifier 를 각 타입내부로 위치 이동

Ldoy committed:
refactor: cell오토레이아웃 오류나오지 않도록 수정

Ldoy committed:
feat: 현재날씨 알려주는 헤더뷰에 데이터 받아와서 화면에 표시하는 로직 구현

- WeatherTableViewDelegate 에서 현재날씨 이미지 로딩
- WeatherTableviewDataSource 이미지로딩 중 필요없은 UIImage 인스턴스 삭제
- WeatherTableViewDelegate에서 이미지로딩 시 사용할 캐시 아이디값으로 date전달하기 위해 CurrentWeather에 date 속성 추가
- ImageLoader 타입 주석 정리
- OpenWeatherHeaderView 에서 iconImageView를 상수로 변경

Ldoy committed:
feat: 테이블뷰에 델리게이트 할당

- 헤더뷰의 높이 지정
- self 캡쳐리스트에 추가
- WeatherDataHolder 에서 마지막 데이터 왔을 때 테이블 뷰 리로드하도록 NoticifationCenter 위치 변경
- PathOptions의 잘못 추가된 .current 케이스의 값 변경
- FiveDaysForecastCell 의 오토레이아웃 값 변경

Ldoy committed:
feat: 온돈바꿔주는 역할 객체 구현 및 cell, view에 적용

Ldoy committed:
feat: 주소를 헤더뷰에 반영하도록 주소를 비동기적으로 받아오는 코드 위치 변경

Ldoy committed:
chore:  주소 변환시 사용하지 않는 extension삭제

Ldoy committed:
faet: 테이블뷰 배경화면 변경하기 위한 UIView에 extension메소드 추가

- tableView와 cell배경색 clear로 변경

Ldoy committed:
feat: refresh control 인스턴스 추가 및 새로고침 기능 구현

Ldoy committed:
chore: 인덴트 정리

Ldoy committed:
refactor: 헤더뷰 레이아웃 수정 및 표시하는 주소 변경

Ldoy committed:
refactor: 그룹위치 변경 및 OpenWeatherMainViewController내부에 있던 CLLocationDelegate 역할 분리

Ldoy committed:
chore: 사용하지 않는 타입 및 주석 삭제
